File: challenge_6/python/shaowen/ranges.py
# -*- coding: utf-8 -*-
"""
Created on Sun Jan  8 09:24:29 2017

@author: Shaowen liu
"""
def ranges(arr):
    """
    para arr, a python list    
    """
    if not arr: return []
    res = []
    # set inital cursor start and end
    start = end = arr[0]
    for i in arr[1:]:
        # if arr in seuquence, update end
        if i == end + 1:
            end = i
        # if arr diverged, reset start & end
        else:
            if start != end:
                res.append('%s->%s'% (start, end))
            start = end = i
    # complete the last range
    if start != end:
        res.append('%s->%s'% (start, end))
    return res

#arr = [1,2,3,4,5,8,9,10]
#print ranges(arr)

File: challenge_6/java/jdfurlan/src/Ranges.java

import java.util.ArrayList;

public class Ranges {

	public static void main(String[] args) {
		int[][] arr = { { 1, 2, 3, 4, 8, 9, 10, 12, 13, 14 }, { 1, 2, 3, 4, 9, 10, 15 }, { 1, 2 }, { 1 },
				{ 1, 2, 3, 4, 5, 8, 9, 10}, { 1, 3 }};
		for(int[] a : arr){
			System.out.println(findRanges(a));
		}
	}

	private static ArrayList<String> findRanges(int[] arr) {
		ArrayList<String> res = new ArrayList<String>();
		int first, last;
		for (int i = 0; i < arr.length - 1; i++) {
			first = arr[i];
			last = 0;
			while (i < arr.length-1 && arr[i + 1] == arr[i] + 1) {
				i++;
				last = arr[i];
			}
			if (last != 0)
				res.add(first + "->" + last);
		}
		return res;
	}

}
